behave_reportportal/behave_agent.py

```diff
@@ -0,0 +1,253 @@
+            return
+        description = step.text or ""
+        self._step_id = self._rp.start_test_item(
+            name="[{keyword}]: {name}".format(
+                keyword=step.keyword, name=step.name
+            ),
+            start_time=timestamp(),
+            item_type="STEP",
+            parent_item_id=self._scenario_id,
+            code_ref=self._code_ref(step),
+            description=description + self._build_table_content(step.table),
+            **kwargs
+        )
+
+    @check_rp_enabled
+    def finish_step(self, context, step, **kwargs):
+        """Finish test in Report Portal."""
+        if self._cfg.step_based:
+            self._finish_step_step_based(step, **kwargs)
+            return
+        self._finish_step_scenario_based(step, **kwargs)
+
+    def _get_launch_attributes(self):
+        """Return launch attributes in the format supported by the rp."""
+        attributes = self._cfg.launch_attributes or []
+        system_attributes = get_launch_sys_attrs()
+        system_attributes["agent"] = "{}-{}".format(
+            self.agent_name, self.agent_version
+        )
+        return attributes + _dict_to_payload(system_attributes)
+
+    @staticmethod
+    def _build_table_content(table):
+        if not table:
+            return ""
+
+        pt = PrettyTable(field_names=table.headings)
+        [pt.add_row(row.cells) for row in table.rows]
+        return "\n" + pt.get_string()
+
+    def _finish_step_step_based(self, step, status=None, **kwargs):
+        if step.status.name == "failed":
+            self._log_step_exception(step, self._step_id)
+        self._rp.finish_test_item(
+            item_id=self._step_id,
+            end_time=timestamp(),
+            status=status or self.convert_to_rp_status(step.status.name),
+            **kwargs
+        )
+
+    def _finish_step_scenario_based(self, step, **kwargs):
+        self._rp.log(
+            item_id=self._scenario_id,
+            time=timestamp(),
+            message="[{keyword}]: {name}. {text}{table}".format(
+                keyword=step.keyword,
+                name=step.name,
+                text=step.text or "",
+                table=self._build_table_content(step.table),
+            ),
+            level="INFO",
+            **kwargs
+        )
+        if step.status.name == "failed":
+            self._log_step_exception(step, self._scenario_id)
+
+    def _log_step_exception(self, step, item_id):
+        if not step.exception:
+            return
+        self._rp.log(
+            item_id=item_id,
+            time=timestamp(),
+            level="ERROR",
+            message="Step [{keyword}]: {name} was finished with "
+            "exception:\n{exception}".format(
+                keyword=step.keyword,
+                name=step.name,
+                exception=", ".join(step.exception.args),
+            ),
+        )
+
+    @staticmethod
+    def _item_description(item):
+        if not item.description:
```

Comment:
Can you please explain why it is more safe? Item is Behave entity with strict structure. And both option return AttributeError in some cases, but using of fields name as strings can lead to typos.

Reply:
From my experience, the framework's objects and their attributes tend to change over time. Check out changes in Robot Framework 4, for example. Also, `object.attribute` syntax does not guarantee the existence of that attribute in the object. You are not using type hints here...

Reply:
Can we support just python 3 and use type hints?
Anyway `getattr(object, "nonexistent_attribute")` also raises AttributeError, but you should update strings instead of object fields

Reply:
Type hints can be used with Python 2.7 as well. Check out *.pyi files in the common client. `getattr(object, "nonexistent_attribute", None)` will not raise.

Reply:
if we will use `getattr(object, "nonexistent_attribute", None)` it will look like empty field and users will not know that contract of library was changed, is it ok?

Reply:
Good point.